### PR TITLE
New endpoint to expose the data for the configurable API

### PR DESCRIPTION
**Pull request by gavriil-g-moj**
This PR exposes the following endpoint:
`GET {reportId}/{dataSetId}/{reportVariantId}?filters.date.start=2023-05-15&filters.date.end=2023-06-20&selectedPage=1&pageSize=10&sortColumn="direction"&sortedAsc=true`

This gets the ProductDefinition based on the reportId (reportId == ProductDefinitionId) and validates the filters and sort column. 
Then calls the ConfiguredApiRepository which constructs a query based on the query of the dataSet of the given ProductDefinition and all the filters, sort column and asc/desc params that have been passed to the controller request.

It provides a configurable way of retrieving report data which is not tied to any hardcoded table or column names.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.RestController
+import uk.gov.justice.digital.hmpps.digitalprisonreportingmi.service.ConfiguredApiService
+
+@Validated
+@RestController
+@Tag(name = "Configured Data API")
+class ConfiguredApiController(val configuredApiService: ConfiguredApiService) {
+
+  @GetMapping("/{reportId}/{dataSetId}/{reportVariantId}")
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:19]: In the definition list endpoint, it currently returns just the dataset ID as the resourceName.

Could you update that to return this pattern instead, please? That way the resourceName the UI uses for API calls will match this endpoint.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @PathVariable("reportVariantId") reportVariantId: String,
+  ): List<Map<String, Any>> {
+    return configuredApiService.validateAndFetchData(reportId, dataSetId, reportVariantId, filtersOnly(allQueryParams), selectedPage, pageSize, sortColumn, sortedAsc)
+  }
+
+  private fun filtersOnly(filters: Map<String, String>): Map<String, String> {
+    return filters.entries
+      .filter { it.key.startsWith("filters.") }
+      .filter { it.value.isNotBlank() }
+      .associate { (k, v) -> k.removePrefix("filters.") to v }
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:45]: Currently the UI doesn't send the "filters." prefix to the API, same as it doesn't send the "paging." prefix that it uses for the pages.

It's not a big deal to add though. In fact, it may simplify it slightly.

Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:45]: For the paging the API doesn't expect any prefix, I only added the "filters." prefix for the filters.
I could either remove it from here or it can be added to the request from the UI. 
I am happy with either way.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt (hunk 1 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:19]: Would this mean to have only the dataSet id as a path parameter and to move the reportId and reportVariantId as query parameters?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        it.entries.associate { (k, v) ->
+          if (v is Timestamp) {
+            k to v.toLocalDateTime().toLocalDate().toString()
+          } else {
+            k to v
+          }
+        }
+      }
+    stopwatch.stop()
+    log.info("Query Execution time in ms: {}", stopwatch.time)
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt:71]: Should this be debug level information?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/StubbedProductDefinitionRepository.kt [hunk 4]
```diff
@@ -220,4 +222,11 @@ class StubbedProductDefinitionRepository : ProductDefinitionRepository {
       ),
     )
   }
+
+  override fun getDataSet(reportId: String, dataSetId: String): DataSet {
+    return getProductDefinitions().filter { it.id == reportId }
+      .flatMap { pd -> pd.dataSet.filter { it.id == dataSetId } }
+      .ifEmpty { throw ValidationException("Invalid dataSetId provided: $dataSetId") }
+      .takeIf { it.size == 1 }?. first() ?: throw DuplicateDataSetIdException(dataSetId)
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/StubbedProductDefinitionRepository.kt:230]: I think once the definition is put into the appropriate data store, it will be the job of whatever adds the definition to validate it. This feels like a redundant validation in that context.

Also, the definition and variant IDs are not validated, which makes this a bit inconsistent.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    query: String,
+    rangeFilters: Map<String, String>,
+    filtersExcludingRange: Map<String, String>,
+    selectedPage: Long,
+    pageSize: Long,
+    sortColumn: String,
+    sortedAsc: Boolean,
+  ): List<Map<String, Any>> {
+    val preparedStatementNamedParams = MapSqlParameterSource()
+    // TODO: Consider case insensitive non range filters
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt:32]: Probably prudent, given the inconsistent casing in some of the data.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        "${k.removeSuffix(".end")} <= :$k"
+      } else {
+        throw ValidationException("Range filter does not have a .start or .end suffix: $k")
+      }
+    }.joinToString(" AND ")
+      .removeSuffix(" AND ")
+      .ifEmpty { null }
+
+//    val whereClause = whereNoRange?.let{"WHERE $it AND $whereRange"} ?: whereRange?.let { "WHERE $it" } ?: ""
+    val whereClause = whereNoRange?.let { "WHERE ${whereRange?.let { "$whereNoRange AND $whereRange"} ?: whereNoRange}" } ?: whereRange?.let { "WHERE $it" } ?: ""
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt:49]: I'm finding this line a bit hard to parse. It might help to decompress it a little, into a few different statements.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt (hunk 3 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt:71]: I have changed it to debug.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/StubbedProductDefinitionRepository.kt (hunk 4 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/StubbedProductDefinitionRepository.kt:230]: I have now removed the validation which throws the DuplicateDataSetIdException.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt (hunk 1 above)
Comment (stalwartsolutions) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:19]: I mean this line could be updated to match, so the UI uses the right "resource": https://github.com/ministryofjustice/hmpps-digital-prison-reporting-mi/blob/main/src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/service/ReportDefinitionMapper.kt#L48

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt (hunk 2 above)
Comment (stalwartsolutions) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:45]: Leave it be and I'll add a matching change in the UI. Would it be worth adding a comment to the map parameter so the Swagger docs show what's expected?

Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt:45]: Thanks, I have added a comment to the API description as it does not allow generic comments in the query parameters.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    val whereRange = rangeFilters.keys.map { k ->
+      if (k.endsWith(".start")) {
+        "${k.removeSuffix(".start")} >= :$k"
+      } else if (k.endsWith(".end")) {
+        "${k.removeSuffix(".end")} <= :$k"
+      } else {
+        throw ValidationException("Range filter does not have a .start or .end suffix: $k")
+      }
+    }.joinToString(" AND ")
+      .removeSuffix(" AND ")
```
[the remark this answers is not in the document]
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt:44]: I have now changed that.

[the remark this answers is not in the document]
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt]: I have now changed that.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt (hunk 5 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/data/ConfiguredApiRepositoryCustom.kt]: Done.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt (hunk 1 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt]: I have now changed this to: `"$productDefinitionId/${report.id}"`

Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportingmi/controller/ConfiguredApiController.kt]: And the path of the ConfiguredApiController has also been changed to `"/{reportId}/{reportVariantId}"`

**Review by stalwartsolutions — APPROVED** [review]

**Review by stalwartsolutions — APPROVED** [review]
